### ISSUE TITLE
How could I use multi-gpu 

### ISSUE THREAD

**Issue by CoinCheung** [VisualComputingInstitute/triplet-reid#58]
It seems that the model only runs on a single gpu no matter how many gpus are available. If the space the model takes up is more than the volume of one gpu, there would be oom error. I  can train the model on a single gpu with default configuration, but once I double the batch size and use two gpus, there is oom errors. How could I use multi-gpu in this case ?

**Comment by Pandoro** [comment]
This sadly doesn't work out of the box in Tensorflow, you will need to adjust the code quite a bit for this to work. For example you could start by taking a look at [this example](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py). This is nothing we are planning on doing though.

**Comment by CoinCheung** [comment]
@chris20181220  yes, I reimplement it with pytorch, and my [implementation](https://github.com/CoinCheung/triplet-reid-pytorch) supports multi-gpu working

**Comment by chris20181220** [comment]
> @chris20181220 yes, I reimplement it with pytorch, and my [implementation](https://github.com/CoinCheung/triplet-reid-pytorch) supports multi-gpu working

@CoinCheung if I need to use tf, do u know how to fix?

**Comment by CoinCheung** [comment]
@chris20181220 As the author said, it will be quite tedious and many code should be modified, I do not think I can do it now. Sorry I cannot help.

**Comment by chris20181220** [comment]
> @chris20181220 As the author said, it will be quite tedious and many code should be modified, I do not think I can do it now. Sorry I cannot help.

@CoinCheung OK thank you all the same, i try to modify

**Comment by lucasb-eyer** [comment]
You should also be aware that there comes the question of how to do the triplet mining in the batch: mine on each GPU's batch independently, or gather all batch outputs to one fixed GPU and mine in the large complete batch there. There are trade-offs and it's not clear what is best.

**Comment by Pandoro** [comment]
Also keep in mind what you do with the batch normalization. When you split
the batch, it could pay off to specifically split the batch to make two
P×K/2 batches, instead of two P/2×K batches, unless you specifically sync
your batch normalization across GPUs.
